parity-clib/parity.h

```diff
@@ -14,166 +14,270 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-#ifndef _PARITY_H_INCLUDED_
-#define _PARITY_H_INCLUDED_
-
+/// The C interface to Parity Ethereum.
+///
+/// # Thread safety
+///
+/// The Parity Ethereum C API is thread safe. All resources can be operated on
+/// by multiple threads simultaneously.
+///
+/// Your callbacks are also expected to be thread safe. Parity Ethereum makes
+/// heavy use of background threads internally, rather than using a
+/// user-provided event loop. Therefore, your callbacks may be called
+/// from any thread, or even from multiple threads at once. They need to be
+/// prepared for that. A good way to handle this is for your callbacks to
+/// deserialize the message and then use a thread-safe queue to deliver the
+/// message to your event loop.
+///
+/// # Blocking
+///
+/// The Parity Ethereum C API generally does not block on network I/O, except
+/// when a Parity Ethereum instance is being destroyed. However, it may block on
+/// disk I/O at any time. If this presents a problem in your application, you
+/// should call the Parity Ethereum C API from a worker thread that is allowed
+/// to block.
+///
+/// Your callbacks should also not block under normal circumstances. They do not
+/// need to be real-time, but you are expected to be able to keep up with the
+/// events you have subscribed to. The blockchain won’t slow down for you, so if
+/// you cannot keep pace with incoming messages, you will fall behind.
+///
+/// # Constructing multiple Parity Ethereum instances
+///
+/// Constructing multiple instances of Parity Ethereum is discouraged. While it
+/// is expected to work, they will duplicate a large amount of state
+#ifndef PARITY_H_INCLUDED
+#define PARITY_H_INCLUDED 1
 #include <stddef.h>
+#include <stdint.h>
+// Rust assumes that size_t and uintptr_t are the same.  Check that that is the
+// case to avoid very obscure errors.
+#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#include <type_traits>
+static_assert(std::is_same<uintptr_t, size_t>::value,
+              "Rust code requires that uintptr_t and size_t are the same");
+#endif
+extern "C" {
+#elif __STDC_VERSION__ >= 201112L
+_Static_assert(sizeof(uintptr_t) == sizeof(size_t),
+               "Rust code requires that uintptr_t and size_t are the same");
+#endif
+
+/// An opaque struct that represents a Parity configuration.
+struct parity_config;
```

Comment:
It is "safe" to use `struct foo;` and `foo*` rather than using `void*`?
From my experience, any attempt to be type-safe in C is a bad idea that generally backfires.

Reply:
It should be, IMO.  `void*` and `struct foo*` have the same ABI on all platforms I know of.

How does it backfire in your experience?

Reply:
I don't have a strong opinion on `void*` vs `struct foo*` so feel free to discard. My question was mostly about whether `void*` is truly equivalent to `struct foo*`.

With `struct foo;` it would for example be possible for some to write `struct foo { int a; };` and try to use `foo->a`. But I would consider that as being in the "don't do that" category, rather than the "safety" category.

As for the backfire: I don't think there's anything wrong with what this PR does. Trying to squeeze oriented-object programming into C (a language that isn't made for that) tends to lead to very weird or annoying APIs (think GTK for example), so we shouldn't try *too hard* to be strongly-typed.

Reply:
Agreed.  IMO, the main use of the C API is for wrappers in higher-level languages.